Fixed #26306 -- Memory leak in cached template loader

AlexHill opened the pull request:
This fixes two related issues in the caching template loader, a memory leak bug and a memory use regression.

In 1.8, the thing that is cached when a template isn't found is the TemplateDoesNotExist class itself. When the debug view wants to know which paths were searched while looking for the template, it consults the rendering engine again. In 1.9, that debug information is stored on an exception instance, and those exceptions are cached.

The memory leak comes from the repeated re-raising of the same cached exception, which results in references to the `TemplateResponse` for each failed request persisting deep in the exception's context/stack information. The fix is to raise a new exception each cache hit, instead of repeatedly raising the cached exception.

The memory use regression comes from the intentional caching of every checked template path for each failed template, for debug purposes. In some projects that can end up eating all your memory. This is fixed by reverting to the 1.8 behaviour – simply caching the `TemplateDoesNotExist` class – when template debugging is not enabled.

charettes commented on django/template/loaders/cached.py:
```diff
@@ -26,12 +26,25 @@ def __init__(self, engine, loaders):
+        elif isinstance(exc, TemplateDoesNotExist):
+            if self.engine.debug:
+                return exc.__class__(*exc.args, tried=exc.tried)
+            return exc.__class__
```
Shouldn't we also include `*exec.args` in this case?

AlexHill replied:
For memory reasons, we want to cache just the raw exception class, so that the overhead for each missing template is just a single reference to an existing object, rather than a unique object for each template.

Ultimately, the exception that propagates upward in the code is [this one](https://github.com/django/django/blob/0e80ac467293c659588769ec0587a5bf5f084b49/django/template/engine.py#L169), and that contains the template name.

charettes replied:
I see, that makes sense!

AlexHill replied:
Just having a look at the test failure there, looks like I'll need to change the approach a little.

We still want to cache the raw request, but `TemplateDoesNotExist` has a required positional argument, so can't be raised "raw".

AlexHill commented:
Ok, reworked this so everything works properly and all tests pass. There's still a failure under 2.7 which I've just fixed – the leaky exception attributes didn't exist before Python 3.

timgraham commented on django/template/backends/django.py:
```diff
@@ -97,13 +97,19 @@ def render(self, context=None, request=None):
             reraise(exc, self.backend)
 
 
+def copy_exception(exc, backend=None):
```
Add docstring?

timgraham commented on django/template/loaders/cached.py:
```diff
@@ -26,12 +27,23 @@ def __init__(self, engine, loaders):
 
+    def _raise_if_exception(self, value, template_name):
+        """
+        Take a value pulled from the template cache, and raise an appropriate
```
chop comma

timgraham commented on django/template/loaders/cached.py:
```diff
@@ -26,12 +27,23 @@ def __init__(self, engine, loaders):
+        Take a value pulled from the template cache, and raise an appropriate
+        exception. If template debugging is off, the cached item will be the
+        raw TemplateDoesNotExist class, otherwise it will be an instance of
+        TemplateDoesNotExist that we need to copy before raising.
```
explain why it needs to be copied

timgraham commented on django/template/loaders/cached.py:
```diff
@@ -26,12 +27,23 @@ def __init__(self, engine, loaders):
         if cached:
-            if isinstance(cached, TemplateDoesNotExist):
-                raise cached
+            self._raise_if_exception(cached, template_name)
```
Not sure that a separate method here has much advantage. What was your thinking about that?

AlexHill replied:
Just to make `get_template` easier to read, and to provide a convenient place to write a substantial comment on that bit of code. I like small functions.

Happy to change this if you prefer.

timgraham commented on django/template/loaders/cached.py:
```diff
@@ -39,7 +51,7 @@ def get_template(self, template_name, template_dirs=None, skip=None):
             )
         except TemplateDoesNotExist as e:
-            self.get_template_cache[key] = e
+            self.get_template_cache[key] = copy_exception(e) if self.engine.debug else TemplateDoesNotExist
```
A comment about this would be useful.

AlexHill replied:
As per my comment above I think I'd like to just move this into its own function and write a substantial docstring. The reasons for caching what I'm caching are much easier to explain in five lines than two, and I don't want to clutter this function with block comments.

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -50,11 +50,30 @@ def test_get_template(self):
         e = self.engine.template_loaders[0].get_template_cache['doesnotexist.html']
         self.assertEqual(e.args[0], 'doesnotexist.html')
 
+        self.engine.debug = False
```
can this be a separate test method?

AlexHill replied:
Yep

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -50,11 +50,30 @@ def test_get_template(self):
+        e = self.engine.template_loaders[0].get_template_cache['alsodoesnotexist.html']
+        self.assertEqual(e, TemplateDoesNotExist)
+
+    def test_cached_exception_no_traceback(self):
```
I think this test is getting into the "not extremely obvious" area, so a docstring would be useful.

AlexHill replied:
Ok

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -50,11 +50,30 @@ def test_get_template(self):
+            self.engine.get_template('cachestemplateresponse.html')
+        e = self.engine.template_loaders[0].get_template_cache['cachestemplateresponse.html']
+
+        if six.PY3:
```
alternatively could use `if six.PY2: return` so we can simply remove the condition instead of also deindenting some lines when dropping Python 2.

AlexHill replied:
I changed this to `@skipIf(six.PY2)` instead.

timgraham commented:
Please add a mention in the 1.9.5 release notes. For future reference, please send the PR against master and let the committer do the backport. No need to do a new PR here at this point through since this merges cleanly to master.

@prestontimmons, care to take a look?

AlexHill commented:
Hi Tim,

Thanks for your feedback. I've incorporated your suggestions into the patch.

The only thing that might be questionable is that I pulled even more code our of `get_template()` into a tiny helper function with a big docstring. I think it all reads better that way but I have no problem changing it at all.

Also rebased against latest `stable/1.9.x`. I'll base my patch on master next time.

Thanks,
Alex

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -49,11 +49,46 @@ def test_get_template(self):
+    @unittest.skipIf(six.PY2, "Python 2 doesn't set extra exception attributes")
+    def test_cached_exception_no_traceback(self):
+        """
+        Ensure the exception cached when template debugging is enabled doesn't
```
Rather than "Ensure..." we are trying to move toward simply stating the expected behavior.

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -49,11 +49,46 @@ def test_get_template(self):
+        """
+        Ensure the exception cached when template debugging is enabled doesn't
+        contain the traceback or context information that Python sets when an
+        exception is raised in Python 3.
```
If you don't mention "Python 3" then that's one less change to make when dropping Python 2.

timgraham commented on django/template/backends/django.py:
```diff
@@ -97,13 +97,24 @@ def render(self, context=None, request=None):
     """
-    Reraise TemplateDoesNotExist while maintaining template debug information.
+    Create a new TemplateDoesNotExist, preserving its declared attributes and
+    template debug data but discarding __traceback__, __context__ and __cause__
```
, and

timgraham commented:
I don't find that the underscore methods make the code any easier to follow. If they were used more than once or meant to be overridden, sure, but otherwise, I see no advantage over inline code/comments.

timgraham commented:
merged in ecb59cc6579402b68ddfd4499bf30edacf5963be, thanks!

mfeliu commented:
Hello @timgraham and @AlexHill . I have a memory leak in 1.8 so I am interested in the backport  of this PR. What is the process to backport it? Should I try to do it myself and generate another PR? Thanks and sorry for the inconvenience.

timgraham commented:
You might be encountering a different issue. As far as I understand, the issues addressed here are new in 1.9.

mfeliu commented:
Ok. Thanks again @timgraham.
